### ISSUE TITLE
Adding index cross-references for definitions where index entry is for a different term

### ISSUE THREAD

**Issue by jwakely** [cplusplus/draft#2067]
We have:

````
A \defnx{trivially copyable class}{class!trivially copyable} is a class:
````

This defines the term _trivially copyable class_ but adds an index subentry for "class, trivially copyable". That means "trivially copyable class" can't be found in the index!

We need to manually add that entry (see #2065 for that change):

````
\indextext{trivially copyable class|see{class, trivially copyable}}%
````

Is there any way to automate that through `\defnx` or some other macro, or at least make it easier? Maybe something like this (although it would need to replace `foo!bar` with `foo, bar` in `#2`):

````
\newcommand{\defnxx}[2]{\indexdefn{#2}\indextext{#1|see{#2}}\textit{#1}}
````

**Comment by godbyk** [comment]
If we redefine `\defnx` as follows:

```latex
\newcommand{\defnx}[2]{%
  \indexdefn{#2}%
  \ifthenelse{\NOT\equal{#1}{#2}}{%
    \indextext{#1|see{#2}}%
  }{}%
  \textit{#1}%
}
```

then `\defnx{trivially copyable class}{class!trivially copyable}` and `\defn{blah}` would both do what you want.

**Comment by jensmaurer** [comment]
Issue #1033 is related; we did decide that the page numbers go with the decomposed index entry. (That doesn't mean we should not have "see" entries for the other form.)
It is not obvious that the "see" form is always non-decomposed. For example "class, trivially copyable" has the page numbers and "trivially copyable, class" has the "see". There might also be a "trivially copyable, type" with another "see" to "type, trivially copyable" that points to [basic.types] or so.
@jwakely, I believe such an approach would still make the index useful to find "trivially copyable class"; do you agree?

**Comment by jwakely** [comment]
> @jwakely, I believe such an approach would still make the index useful to find "trivially copyable class"; do you agree?

Yes. As long as I can go to T in the index and find "trivially copyable class" in some form or another I think that's fine. Currently it's not there at all under T.

I tried using additional packages so I could replace the "|" with ", " in the index entry, but didn't get it working. Breaking it into two arguments seems like a better solution.

**Comment by jensmaurer** [comment]
In #2080, there is a specific change to create a \defnc for compound defined terms with uniform indexing.

**Comment by godbyk** [comment]
@jensmaurer Why are we avoiding `\ifthenelse`?

The definition of `\defnc` in #2080 seems rather narrow in that it only works for _x y → y, x_ transforms. Wouldn't a more general function (like what my `\defnx` macro above provides) be better? Using `\defnc` also means that authors and editors have to learn yet another macro.

**Comment by jensmaurer** [comment]
@godbyk: The point is not the `\ifthenelse`. I'm not seeing that your \defnx macro above converts an exclamation mark to a comma, as required to from a proper "see".

**Comment by godbyk** [comment]
@jensmaurer Ah, sorry. My brain was still in `xindy` mode when I wrote it. (`xindy` requires you use `x!y` instead of `x, y` for cross-references.)

Try replacing the existing `\defnx` macro with the following code:

```latex
\usepackage{xstring}
\newcommand{\defnx}[2]{%
  \indexdefn{#2}%
  \ifthenelse{\NOT\equal{#1}{#2}}{%
    \StrSubstitute{#2}{!}{, }[\stdtemp]%
    \indextext{#1|see{\stdtemp}}%
  }{}%
  \textit{#1}%
}
```

It replaces all instances of `!` with `, ` in the index entry so the cross-reference should work again. (If we switch to `xindy` in the future, we'll want to skip the substitution, though.)

**Comment by jensmaurer** [comment]
@godbyk: Ah. So, we get the correct "see" when using exclamation marks once xindy lands (see #1917). Thanks.
I'd like to point out, though, that the `\defnc` macro reduces redundancy in the source code, because every word appears only once. (A `\defnx` repeats the words in a different order.) Although it's awkward that the position of the exclamation mark is indicated by `}{` (i.e. starting the second argument).

**Comment by godbyk** [comment]
@jensmaurer Yeah. xindy actually checks to see if the cross-reference target exists and complains if it doesn't. It's found a few errors in our current indexing.

It's true that it reduces the number of words you have to type, but only for that particular construction. If you want to do any other type of cross-reference, you're stuck using `\defnx` anyway. I think using `\defnx` in all cases will reduce the mental load on authors: it provides consistency and clarity.

**Comment by jensmaurer** [comment]
@godbyk Right, but I suspect we'll have a lot of use for `\defnc` once we actually get going with #1033. Let's have the editorial team discuss this in Rapperswil.

**Comment by jensmaurer** [comment]
Editorial meeting consensus: \defnc as presented in #2080 is approved and even liked by the Project Editor. Yes, it covers only two-part terms, but that's likely > 95% of the cases, so good enough.

**Comment by jensmaurer** [comment]
"trivially copyable class" was fixed. There's no point in leaving this open. Please create a new issue for specific bad index entries.
